### DISCUSSION TITLE
運用旗標寫防彈跳機制

### DISCUSSION THREAD

**Discussion by YOGA012** [1101-GS4719-C/Python-Community-GS4719-C#1969] [General]
大一時，我修了系上開設的微控制器實驗，使用的程式語言是C，當時做到用數字鍵盤輸入訊號給ASA匯流排微電腦-M128主機板(後面簡稱ASA板)，然後再經由ASA板輸出訊號給七節管顯示鍵盤按的數字，實驗當時用了四個七節管，要求的成果是按下數字鍵盤要能夠顯示能夠顯示數字，並且按下下一個數字時，前一個數字要能夠自動進位到下一個七節管。

我將程式碼分為三個函式，分別為主函式、讀取數字鍵盤訊號以及輸出訊號至七節管的兩個副函式。最一開始，我的構想是在主函式寫一個while無限迴圈，利用if條件句判斷是否有輸入訊號，一次按下的動作視為一次訊號，如果ASA板讀取到輸入訊號就會輸出對應的數字訊號給七節管。雖然程式可以執行，不過若是在鍵盤被按下的時間內，迴圈有重複執行，那就會重複輸入及輸出訊號，有點像用電腦打字時長按a就會跑出一堆aaaaaaaaaaa，也就是說目前的寫法變成一次迴圈做一次讀取輸入及輸出，而不是原先設定一次按下動作為一次訊號，因此勢必要對程式做一些修改。由於副函式主要功能為輸入及輸出，只要能夠正常讀取及輸出訊號即可，所以要修改的只有主函式。

為了解決這個問題就需要在主函式另外寫一個防彈跳措施，其目的就是防止一次按下的動作被多次讀取，為了完成這個防彈跳措施，需要引入一個重要的觀念「**旗標**」。以下是我原本的程式碼以及修改後的程式碼。
修改前:
```
int main(){
    uint8_t key;
    while(1){
        key=get()    #以key變數作為讀取鍵盤的判斷，get函式是讀取鍵盤按鍵的副函式，key=1為有訊號，key=2為無訊號
        if(key=1){
            輸出訊號至七節管;
        }
        _delay_ms(10);#延遲10毫秒後再做一次讀取
    }
}
```

修改後:
```
int main(){
    uint8_t key;
    int flag = 1;    #先將旗標設為1
    while(1){
        key=get()    #以key變數作為讀取鍵盤的判斷，get函式是讀取鍵盤按鍵的副函式，key=1為有訊號，key=2為無訊號
        if(key=1){
            if(flag=1){
                輸出訊號至七節管;
            }
            flag = 0;
        }
        else{
            flag = 1;
        }
        _delay_ms(10);#延遲10毫秒後再做一次讀取
    }
}
```
由修改後的程式碼可以看出，我將原本while迴圈內只做一次if判斷是否讀到輸入訊號的條件更改為需要再做一次if判斷flag是否為1才能做訊號輸出，並且輸出完訊號後就將flag=0，直到放開鍵盤後進入else才將flag變回1。如此一來，就算一直按著鍵盤不放，ASA板也確實有讀到輸入訊號(key為1)，但因為flag已經變成0而無法輸出訊號，藉此來達成防彈跳機制。

**Comment by jeff110303549** [comment]
你對程式碼的修改很有見解，尤其是引入了防彈跳的概念以解決重複輸入的問題。這種在主函式中設置旗標的方法確實可以有效地控制輸入和輸出的流程，使程式更加穩定和可靠。
通過你的描述和程式碼，我能感受到你對解決問題的熱情和技巧，這對於解決實際應用中遇到的各種挑戰將會非常有幫助。继续保持好奇心和創造力，這對於你的學習和成長都會有很大的助益。

**Comment by Piyenparty** [comment]
它雖然看起來只是個簡單的邏輯旗標，但在處理輸入（尤其是硬體或重複觸發場景）時，卻能大幅提高穩定性與使用者體驗。

其實很多程式問題不是出在功能沒寫，而是在時機點、邏輯控制或異常處理上。能夠把「輸入控制」這類看似不起眼的細節做好，反而是讓整體系統更穩的關鍵。若不是你的分享，我這輩子都不一定有機會接觸到這相關的知識，真的很謝謝你
